main-command/src/main/scala/sbt/internal/client/NetworkClient.scala

```diff
@@ -190,82 +378,497 @@ class NetworkClient(configuration: xsbti.AppConfiguration, arguments: List[Strin
+    if (args.exists(_.startsWith(NetworkClient.completions)))
+      System.exit(complete(args, useJNI, System.in, System.out))
+    else {
+      val hook = new Thread(() => {
+        System.out.print(ConsoleAppender.ClearScreenAfterCursor)
+        System.out.flush()
+      })
+      Runtime.getRuntime.addShutdownHook(hook)
+      System.exit(Terminal.withStreams {
+        try client(args, System.in, System.err, System.out, useJNI)
+        finally {
+          Runtime.getRuntime.removeShutdownHook(hook)
+          hook.run()
+        }
+      })
     }
   }
 }
 
 object NetworkClient {
+  private[client] class Arguments(
+      val baseDirectory: File,
+      val sbtArguments: Seq[String],
+      val commandArguments: Seq[String],
+      val completionArguments: Seq[String],
+      val sbtScript: String,
+  ) {
+    def withBaseDirectory(file: File): Arguments =
+      new Arguments(file, sbtArguments, commandArguments, completionArguments, sbtScript)
+  }
+  private[client] val completions = "--completions"
+  private[client] val noTab = "--no-tab"
+  private[client] val noStdErr = "--no-stderr"
+  private[client] val sbtBase = "--sbt-base-directory"
+  private[client] def parseArgs(args: Array[String]): Arguments = {
+    var sbtScript = if (Properties.isWin) "sbt.cmd" else "sbt"
+    val commandArgs = new mutable.ArrayBuffer[String]
+    val sbtArguments = new mutable.ArrayBuffer[String]
+    val completionArguments = new mutable.ArrayBuffer[String]
+    val SysProp = "-D([^=]+)=(.*)".r
+    var base: File = new File("")
+    val sanitized = args.flatMap {
+      case a if a.startsWith("\"") => Array(a)
+      case a                       => a.split(" ")
+    }
+    var foundCompletions = false
+    var i = 0
+    while (i < sanitized.length) {
+      sanitized(i) match {
+        case a if foundCompletions => completionArguments += a
+        case a if a == noStdErr || a == noTab || a.startsWith(completions) =>
+          foundCompletions = true
+          completionArguments += a
+        case a if a.startsWith("--sbt-script=") =>
+          sbtScript = a.split("--sbt-script=").lastOption.getOrElse(sbtScript)
+        case a if a.startsWith("--sbt-base-directory=") =>
+          base = a.split("--sbt-base-directory=").lastOption.map(file).getOrElse(base)
+        case a if !a.startsWith("-") && !foundCompletions => commandArgs += a
+        case a @ SysProp(key, value) =>
+          System.setProperty(key, value)
+          sbtArguments += a
+        case a if !foundCompletions =>
+          sbtArguments += a
+      }
+      i += 1
+    }
+    new Arguments(base.getCanonicalFile, sbtArguments, commandArgs, completionArguments, sbtScript)
+  }
   def run(configuration: xsbti.AppConfiguration, arguments: List[String]): Unit =
     try {
-      new NetworkClient(configuration, arguments)
+      val client = new NetworkClient(configuration, parseArgs(arguments.toArray))
+      System.exit(try {
```

Comment:
Either don't system exit here or use `sys.exit` and change the result type to `Nothing` and fix the errors that follow:

```
[error] ${BASE}/main-command/src/main/scala/sbt/BasicCommands.scala:414:22: dead code following this construct
[error]     NetworkClient.run(s0.configuration, arguments)
[error]                      ^
[error] one error found
[error] (commandProj / Compile / compileIncremental) Compilation failed
```

(There's also a `NetworkClient.run` call `xMain`, but `mainProj` never starts to compile.)

Reply:
That's a good point about not calling System.exit. I am curious though why you are seeing those errors because I have never had issues compiling the project with these changes and CI works. Is there anything unusual about your build environment?

Reply:
Those are the errors that follow from changing to sys.exit and recompiling.

Reply:
Aah. Reading comprehension fail.

Reply:
I removed the `System.exit` in `NetworkClient.run` among other changes.